chalice/package.py

```diff
@@ -247,22 +247,183 @@ def _inject_restapi_outputs(self, template):
+
+        return {
+            'Type': 'AWS::ApiGatewayV2::Route',
+            'Properties': {
+                'ApiId': api_ref,
+                'RouteKey': route_key,
+                'Target': {
+                    'Fn::Join': [
+                        '/',
+                        [
+                            'integrations',
+                            {'Ref': integration_ref},
+                        ]
+                    ]
+                },
+            },
+        }
+
+    def _generate_websocketapi(self, resource, template):
+        # type: (models.WebsocketAPI, Dict[str, Any]) -> None
+        resources = template['Resources']
+        api_ref = {'Ref': 'WebsocketAPI'}
+        resources['WebsocketAPI'] = {
+            'Type': 'AWS::ApiGatewayV2::Api',
+            'Properties': {
+                'Name': resource.name,
+                'RouteSelectionExpression': '$request.body.action',
+                'ProtocolType': 'WEBSOCKET',
+            }
+        }
+
+        self._add_websocket_lambda_integrations(api_ref, resources)
+
+        route_key_names = []
+        for route in resource.routes:
+            key_name = 'Websocket%sRoute' % route.replace(
+                '$', '').replace('default', 'message').capitalize()
+            route_key_names.append(key_name)
+            resources[key_name] = self._create_route_for_key(route, api_ref)
+
+        resources['WebsocketAPIDeployment'] = {
+            'Type': 'AWS::ApiGatewayV2::Deployment',
+            'DependsOn': route_key_names,
+            'Properties': {
+                'ApiId': api_ref,
+            }
+        }
+
+        resources['WebsocketAPIStage'] = {
+            'Type': 'AWS::ApiGatewayV2::Stage',
+            'Properties': {
+                'ApiId': api_ref,
+                'DeploymentId': {'Ref': 'WebsocketAPIDeployment'},
+                'StageName': resource.api_gateway_stage,
+            }
+        }
+
+        self._inject_websocketapi_outputs(template)
+
+    def _inject_websocketapi_outputs(self, template):
+        # type: (Dict[str, Any]) -> None
+        # The 'Outputs' of the SAM template are considered
+        # part of the public API of chalice and therefore
+        # need to maintain backwards compatibility.  This
+        # method uses the same output key names as the old
+        # deployer.
+        # For now, we aren't adding any of the new resources
+        # to the Outputs section until we can figure out
+        # a consist naming scheme.  Ideally we don't use
+        # the autogen'd names that contain the md5 suffixes.
+        stage_name = template['Resources']['WebsocketAPIStage'][
+            'Properties']['StageName']
+        outputs = template['Outputs']
+        resources = template['Resources']
+        outputs['WebsocketAPIId'] = {
+            'Value': {'Ref': 'WebsocketAPI'}
+        }
+        if 'WebsocketConnect' in resources:
+            outputs['WebsocketConnectHandlerArn'] = {
```

Comment:
minus the web socket url, are these additional outputs useful for consumers?, else it feels like its adding to the public interface outputs (ie. adding compatibility constraints) without clear value

Reply:
Since we do not have any sort of useable build hooks there isn't a way to modify our generated cfn template from within chalice. So the canonical way, or at least the way I do it, is to have some wrpper script that calls `chalice package tmp/something` and then reads in the generated template and modifies it. This means the keys for these things need to be stable, and are implicitly a part to for API contract. We have had to change names before to make lambda function names not change, so this is something other people are doing as well. Here I'm making it more explicit. If its a part of the output we cant change it. Well, it is experimental so we still could, but the point is I'm trying to make a stronger statement about these being a part of the external API so noone accidentally changes them, or misunderstands.